File: PR2/PR2/ExtendedEuclidean.cpp
#include "ExtendedEuclidean.h"

int32_t ExtendedEuclidean(int32_t firstNum, int32_t secondNum, bool output) {

	if (Gcd(firstNum, secondNum) != 1)
		throw "Вводимые числа не взаимно простые \n";

	vector<int32_t> firstStr = { firstNum, 1, 0 };
	vector<int32_t> secondStr = { secondNum, 0, 1 };
	int32_t quotient = 0;

	if (output) {
		cout << "r-------x-------y--------------------\n";
		cout << firstStr[0] << "\t" << firstStr[1] << "\t" << firstStr[2] << endl;
		cout << secondStr[0] << "\t" << secondStr[1] << "\t" << secondStr[2] << endl;
	}

	while (secondStr[0] != 0) {
		quotient = firstStr[0] / secondStr[0];
		vector<int32_t> thirdStr = { firstStr[0] % secondStr[0], firstStr[1] - quotient * secondStr[1], firstStr[2] - quotient * secondStr[2] };

		if (output)
			cout << thirdStr[0] << "\t" << thirdStr[1] << "\t" << thirdStr[2] << "\t" << "quotient = " << quotient << endl;

		firstStr = secondStr;
		secondStr = thirdStr;
	}
	return firstStr[1];
}

File: PR2/PR2/python/extended_euclidean.py
from modulo_comparison import gcd

def extended_euclidean(first_num, second_num, output=False):
    """
    Реализует расширенный алгоритм Евклида для нахождения модульного обратного.
    Принимает два числа (firstNum, secondNum) и флаг для вывода промежуточных шагов.
    """
    if gcd(first_num, second_num) != 1:
        raise ValueError("Вводимые числа не взаимно простые")

    first_str = [first_num, 1, 0]
    second_str = [second_num, 0, 1]

    if output:
        print("r-------x-------y--------------------")
        print(f"{first_str[0]}\t{first_str[1]}\t{first_str[2]}")
        print(f"{second_str[0]}\t{second_str[1]}\t{second_str[2]}")

    while second_str[0] != 0:
        quotient = first_str[0] // second_str[0]
        third_str = [
            first_str[0] % second_str[0],
            first_str[1] - quotient * second_str[1],
            first_str[2] - quotient * second_str[2]
        ]

        if output:
            print(f"{third_str[0]}\t{third_str[1]}\t{third_str[2]}\tquotient = {quotient}")

        first_str = second_str
        second_str = third_str
        
    return first_str[1]